File: Python/2-Medium/00017._电话号码的字母组合/00017._电话号码的字母组合.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        """
        根据输入的数字字符串，返回所有可能的字母组合。

        Args:
        digits: str - 一个仅包含数字2-9的字符串。

        Returns:
        List[str] - 可能的字母组合列表。

        示例:
        >>> solution = Solution()
        >>> solution.letterCombinations("23")
        ['ad', 'ae', 'af', 'bd', 'be', 'bf', 'cd', 'ce', 'cf']
        """
        # 数字到字母的映射
        phone_map = {
            "2": "abc", "3": "def", "4": "ghi", "5": "jkl",
            "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"
        }

        def backtrack(combination: str, next_digits: str) -> None:
            """使用回溯法生成所有可能的字母组合"""
            # 如果没有更多的数字需要检查
            if len(next_digits) == 0:
                # 已经处理完所有数字，当前组合是最终结果之一
                combinations.append(combination)
            else:
                # 遍历当前数字对应的所有可能的字母
                for letter in phone_map[next_digits[0]]:
                    # 递归调用，将当前字母加到已有组合上，并处理下一个数字
                    backtrack(combination + letter, next_digits[1:])

        combinations = []
        if digits:
            backtrack("", digits)
        return combinations

# 实例化Solution类，并测试功能
solution = Solution()
test_digits = "23"
print(f"输入数字字符串: {test_digits}")
print("可能的字母组合:", solution.letterCombinations(test_digits))

File: C++/2-Medium/00017._电话号码的字母组合/00017._电话号码的字母组合.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};
        vector<string> result;
        vector<string> mapping{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        string current;
        backtrack(result, digits, 0, current, mapping);
        return result;
    }

    void backtrack(vector<string>& result, const string& digits, int index, string& current, const vector<string>& mapping) {
        if (index == digits.size()) {
            result.push_back(current);
            return;
        }
        
        string letters = mapping[digits[index] - '0'];
        for (char letter : letters) {
            current.push_back(letter); // 添加当前字母
            backtrack(result, digits, index + 1, current, mapping); // 递归
            current.pop_back(); // 回溯
        }
    }
};

int main() {
    Solution solution;
    string digits = "23";
    vector<string> combinations = solution.letterCombinations(digits);
    for (const string& combination : combinations) {
        cout << combination << endl;
    }
    return 0;
}